src/inference/dreamEubo.ad.js

```diff
@@ -0,0 +1,181 @@
+    }
+
+    // A list of records, reach contains a completed trace
+    // and a list of observed values from a previous execution wpplFn.
+    // Records can be created by using dreamSample.
+    this.records = this.opts.records;
+    this.currRecord;
+
+    this.params = params;
+    this.cont = cont;
+
+    this.wpplFn = wpplFn;
+    this.s = s;
+    this.a = a;
+
+    // true when inside mapData (i.e. local model)
+    // false otherwise (i.e. global model)
+    this.insideMapData = false;
+
+    this.coroutine = env.coroutine;
+    env.coroutine = this;
+  }
+
+  function checkScoreIsFinite(score, source) {
+    var _score = ad.value(score);
+    if (!isFinite(_score)) { // Also catches NaN.
+      var msg = 'DREAM: The score of the previous sample under the ' +
+          source + ' program was ' + _score + '.';
+      if (_.isNaN(_score)) {
+        msg += ' Reducing the step size may help.';
+      }
+      throw new Error(msg);
+    }
+  };
+
+  DREAM.prototype = {
+
+    // Computes the estimation of the objective EUBO by averaging the
+    // gradients for all the records provided to the module through params,
+    // so as to get an estimated expectation.
+    run: function() {
+
+      var dream = 0;
+      var grad = {};
+
+      return util.cpsForEach(
+
+          // Body.
+          function(record, i, records, next) {
+            return this.estimateGradient(record, function(g, dream_i) {
+              paramStruct.addEq(grad, g); // Accumulates gradient estimates.
+              dream += dream_i;
+              return next();
+            });
+          }.bind(this),
+
+          // Continuation.
+          function() {
+            paramStruct.divEq(grad, this.records.length);
+            dream /= this.records.length; // Averages estimations.
+            env.coroutine = this.coroutine;
+            return this.cont(grad, dream);
+          }.bind(this),
+
+          this.records);
+
+    },
+
+    // Computes a single sample estimate of the gradient,
+    // which in case of EUBO is the minus guide scoring on
+    // the target recorded samples.
+    estimateGradient: function(record, cont) {
+
+      // Makes record available throughout the module.
+      this.currRecord = record;
+
+      // paramsSeen tracks the AD nodes of all parameters seen during
+      // a single execution. These are the parameters for which
+      // gradients will be computed.
+      this.paramsSeen = {};
+      this.logq = 0;
+
+      return this.wpplFn(_.clone(this.s), function() {
+
+        var objective = -this.logq;
+
+        // TODO: not sure whether it's needed or not in our case
+        if (ad.isLifted(objective)) { // Handles programs with zero random choices.
+          objective.backprop();
+        }
+
+        var grads = _.mapObject(this.paramsSeen, function(params) {
+          return params.map(ad.derivative);
+        });
+
+        return cont(grads, -ad.value(objective));
```

Comment:
Yea I agree. I initially thought only about the `logq` part which is relevant to the gradients but indeed the objective should also contain `logp`. Will fix.